File: csharp/ShrewdEvolver.UnitTests/LogNavigatorTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using Xunit;

namespace AaronicSubstances.ShrewdEvolver.UnitTests
{
    public class LogNavigatorTest
    {
        class LogPositionHolderImpl
        {
            public LogPositionHolderImpl(string positionId)
            {
                PositionId = positionId;
            }

            public string PositionId { get; }
        }

        [Fact]
        public void TestEmptiness()
        {
            var instance = new LogNavigator<LogPositionHolderImpl>(new List<LogPositionHolderImpl>());
            Assert.False(instance.HasNext());
            Assert.Equal(0, instance.NextIndex);
        }

        [Fact]
        public void TestNext()
        {
            var testLogs = new List<LogPositionHolderImpl>
            {
                new LogPositionHolderImpl("a"),
                new LogPositionHolderImpl("b"),
                new LogPositionHolderImpl("c"),
                new LogPositionHolderImpl("c")
            };
            var instance = new LogNavigator<LogPositionHolderImpl>(testLogs);

            for (int i = 0; i < testLogs.Count; i++)
            {
                Assert.True(instance.HasNext());
                Assert.Equal(i, instance.NextIndex);
                Assert.Equal(testLogs[i], instance.Next());
            }

            Assert.False(instance.HasNext());
            Assert.Equal(4, instance.NextIndex);
        }

        [Fact]
        public void TestNextWithSearchIds()
        {
            var testLogs = new List<LogPositionHolderImpl>{
                new LogPositionHolderImpl("a"),
                new LogPositionHolderImpl("b"),
                new LogPositionHolderImpl("c"),
                new LogPositionHolderImpl("c")
            };
            var instance = new LogNavigator<LogPositionHolderImpl>(testLogs);

            Assert.True(instance.HasNext());
            Assert.Equal(0, instance.NextIndex);
            Assert.Null(instance.Next(x => x.PositionId == "d"));

            Assert.True(instance.HasNext());
            Assert.Equal(0, instance.NextIndex);
            Assert.Equal(testLogs[1], instance.Next(x => new List<string> { "c", "b" }.Contains(x.PositionId)));

            Assert.True(instance.HasNext());
            Assert.Equal(2, instance.NextIndex);
            Assert.Equal(testLogs[2], instance.Next(x => x.PositionId == "c"));

            Assert.True(instance.HasNext());
            Assert.Equal(3, instance.NextIndex);
            Assert.Null(instance.Next(x => x.PositionId == "a"));

            Assert.True(instance.HasNext());
            Assert.Equal(3, instance.NextIndex);
            Assert.Equal(testLogs[3], instance.Next(x => x.PositionId == "c"));

            Assert.False(instance.HasNext());
            Assert.Equal(4, instance.NextIndex);
        }

        [Fact]
        public void TestNextWithSearchAndLimitIds()
        {
            var testLogs = new List<LogPositionHolderImpl>
            {
                new LogPositionHolderImpl("a"),
                new LogPositionHolderImpl("b"),
                new LogPositionHolderImpl("c"),
                new LogPositionHolderImpl("c")
            };
            var instance = new LogNavigator<LogPositionHolderImpl>(testLogs);

            Assert.True(instance.HasNext());
            Assert.Equal(0, instance.NextIndex);
            Assert.Null(instance.Next(x => x.PositionId == "d", null));

            Assert.True(instance.HasNext());
            Assert.Equal(0, instance.NextIndex);
            Assert.Equal(testLogs[0], instance.Next(x => x.PositionId == "a", x => x.PositionId == "b"));

            Assert.True(instance.HasNext());
            Assert.Equal(1, instance.NextIndex);
            Assert.Null(instance.Next(x => x.PositionId == "c", x => x.PositionId == "b"));

            Assert.True(instance.HasNext());
            Assert.Equal(1, instance.NextIndex);
            Assert.Null(instance.Next(x => x.PositionId == "b", x => x.PositionId == "b"));

            Assert.True(instance.HasNext());
            Assert.Equal(1, instance.NextIndex);
            Assert.Equal(testLogs[1], instance.Next(x => x.PositionId == "b", x => x.PositionId == "c"));

            Assert.True(instance.HasNext());
            Assert.Equal(2, instance.NextIndex);
            Assert.Null(instance.Next(x => x.PositionId == "c", x => x.PositionId == "c"));

            Assert.True(instance.HasNext());
            Assert.Equal(2, instance.NextIndex);
            Assert.Equal(testLogs[2], instance.Next(x => x.PositionId == "c", null));

            Assert.True(instance.HasNext());
            Assert.Equal(3, instance.NextIndex);
        }
    }
}

File: java/src/test/java/com/aaronicsubstances/shrewd/evolver/LogNavigatorTest.java
package com.aaronicsubstances.shrewd.evolver;

import static org.testng.Assert.*;

import java.util.Arrays;
import java.util.List;
import org.testng.annotations.Test;

public class LogNavigatorTest {
    
    static class LogPositionHolderImpl {
        private final String positionId;
    
        public LogPositionHolderImpl(String positionId) {
            this.positionId = positionId;
        }
    
        public String getPositionId() {
            return positionId;
        }
    }

    @Test
    public void testEmptiness() {
        LogNavigator<LogPositionHolderImpl> instance = new LogNavigator<>(Arrays.asList());
        assertEquals(instance.hasNext(), false);
        assertEquals(instance.nextIndex(), 0);
    }

    @Test
    public  void testNext() {
        List<LogPositionHolderImpl> testLogs = Arrays.asList(new LogPositionHolderImpl("a"), 
            new LogPositionHolderImpl("b"), new LogPositionHolderImpl("c"), 
            new LogPositionHolderImpl("c"));
        LogNavigator<LogPositionHolderImpl> instance = new LogNavigator<>(testLogs);
        
        for (int i = 0; i < testLogs.size(); i++) {
            assertEquals(instance.hasNext(), true);
            assertEquals(instance.nextIndex(), i);
            assertEquals(instance.next(), testLogs.get(i));
        }
        
        assertEquals(instance.hasNext(), false);
        assertEquals(instance.nextIndex(), 4);
    }
    
    @Test
    public void testNextWithSearchIds() {
        List<LogPositionHolderImpl> testLogs = Arrays.asList(new LogPositionHolderImpl("a"), 
            new LogPositionHolderImpl("b"), new LogPositionHolderImpl("c"),
            new LogPositionHolderImpl("c"));
        LogNavigator<LogPositionHolderImpl> instance = new LogNavigator<>(testLogs);
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 0);
        assertEquals(instance.next(x -> x.getPositionId() == "d"), null);
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 0);
        assertEquals(instance.next(x -> Arrays.asList("c", "b").contains(x.getPositionId())), testLogs.get(1));
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 2);
        assertEquals(instance.next(x -> x.getPositionId() == "c"), testLogs.get(2));
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 3);
        assertEquals(instance.next(x -> x.getPositionId() == "a"), null);
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 3);
        assertEquals(instance.next(x -> x.getPositionId() == "c"), testLogs.get(3));
        
        assertEquals(instance.hasNext(), false);
        assertEquals(instance.nextIndex(), 4);
    }
    
    @Test
    public void testNextWithSearchAndLimitIds() {
        List<LogPositionHolderImpl> testLogs = Arrays.asList(new LogPositionHolderImpl("a"),
            new LogPositionHolderImpl("b"), new LogPositionHolderImpl("c"), 
            new LogPositionHolderImpl("c"));
        LogNavigator<LogPositionHolderImpl> instance = new LogNavigator<>(testLogs);
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 0);
        assertEquals(instance.next(x -> x.getPositionId() == "d", null), null);
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 0);
        assertEquals(instance.next(x -> x.getPositionId() == "a", x -> x.getPositionId() == "b"), testLogs.get(0));
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 1);
        assertEquals(instance.next(x -> x.getPositionId() == "c", x -> x.getPositionId() == "b"), null);
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 1);
        assertEquals(instance.next(x -> x.getPositionId() == "b", x -> x.getPositionId() == "b"), null);
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 1);
        assertEquals(instance.next(x -> x.getPositionId() == "b", x -> x.getPositionId() == "c"), testLogs.get(1));
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 2);
        assertEquals(instance.next(x -> x.getPositionId() == "c", x -> x.getPositionId() == "c"), null);
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 2);
        assertEquals(instance.next(x -> x.getPositionId() == "c", null), testLogs.get(2));
        
        assertEquals(instance.hasNext(), true);
        assertEquals(instance.nextIndex(), 3);
    }
}